src/sync/PostSyncProcessor.js

```diff
@@ -0,0 +1,157 @@
+ */
+
+import { getNextNumber, NUMBER_SEQUENCE_KEYS } from '../database/utilities';
+
+const { REQUISITION_SERIAL_NUMBER, SUPPLIER_INVOICE_NUMBER } = NUMBER_SEQUENCE_KEYS;
+
+export class PostSyncProcessor {
+  constructor(database) {
+    this.database = database;
+    this.recordQueue = new Map(); // Map of [recordId, recordType]
+    this.actionQueue = [];
+    this.checkTables = this.checkTables.bind(this);
+    this.processRecordQueue = this.processRecordQueue.bind(this);
+    this.onDatabaseEvent = this.onDatabaseEvent.bind(this);
+    this.runActionQueue = this.runActionQueue.bind(this);
+    this.delegateByRecordType = this.delegateByRecordType.bind(this);
+    this.postProcessRequisition = this.postProcessRequisition.bind(this);
+    this.postProcessTransaction = this.postProcessTransaction.bind(this);
+
+    this.database.addListener(this.onDatabaseEvent);
+  }
+
+  /**
+ * Respond to a database change event. Must be called from within a database
+ * write transaction.
+ * @param  {string} changeType  The type of database change, e.g. CREATE, UPDATE, DELETE
+ * @param  {string} recordType  The type of record changed (from database schema)
+ * @param  {object} record      The record changed
+ * @param  {string} causedBy    The cause of this database event, either 'sync' or undefined
+ * @return {none}
+ * Runs all the post sync functions triggered by sync events on database through
+ * this.onDatabaseEvent
+ */
+  onDatabaseEvent(changeType, recordType, record, causedBy) {
+    // Exit if not a change caused by incoming sync
+    if (causedBy !== 'sync' || recordType === 'SyncOut') return;
+    if (this.recordQueue.has(record.id)) {
+      // Check if already in queue, remove old
+      this.recordQueue.delete(record.id);
+    }
+    // Add new entry at end of Map
+    this.recordQueue.set(record.id, recordType);
+  }
+
+  /**
+   * Runs post checks across tables, ensuring data is correct.
+   * Tables manually added as to not iterate over tables that don't have any post processing.
+   */
+  checkTables() {
+    this.actionQueue = [];
+    this.recordQueue = []; // Reset the recordQueue to avoid unnessary runs
+
+    this.database
+      .objects('Requisition')
+      .forEach(record => this.delegateByRecordType('Requisition', record));
+
+    this.database
+      .objects('Transaction')
+      .forEach(record => this.delegateByRecordType('Transaction', record));
+
+    this.runActionQueue();
+  }
+
+  /**
+   * Iterates through records added through listening to sync, adding needed actions
+   * to actionQueue. Runs the actionQueue, making the changes.
+   */
+  processRecordQueue() {
+    this.recordQueue.forEach((recordType, recordId) => {
+      // Use local database record, not what comes in sync. Ensures that records are
+      // integrated information (definitely after sync is done)
+      const internalRecord = this.database.objects(recordType).filtered('id == $0', recordId)[0];
```

Comment:
My belief is that in a case where you should always expect a condition to be met, you shouldn't safety check for that condition. If you do, there may be something going wrong that is silenced by your safety check. If you don't, the underlying problem is very brutally and obviously exposed by a crash (and a bugsnag report! heh)